src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,276 @@
+   end
+   n_cols = sum([ncols(V[j]) for j in 1:n])
+   B = zero_matrix(base_ring(V[1]), n_rows, n_cols)
+   pos_i=1
+   for i in 1:m
+      pos_j=1
+      for j in 1:n
+         insert_block!(B,V[n*(i-1)+j],pos_i,pos_j)
+         pos_j += ncols(V[n*(i-1)+j])
+      end
+      pos_i += nrows(V[n*(i-1)+1])
+   end
+   return B
+end
+
+"""
+    matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1})
+Return the matrix whose rows are the vectors in `A`. Of course, vectors in `A` must have the same length and the same base ring.
+"""
+function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <: FieldElem
+   c = length(A[1].v)
+   X = zero_matrix(base_ring(A[1]), length(A), c)
+   for i in 1:length(A)
+      @assert length(A[i].v)==c "Vectors must have the same length"
+      for j in 1:c X[i,j] = A[i][j] end
+   end
+
+   return X
+end
+
+"""
+    conjugate_transpose(x::MatElem{T}) where T <: FieldElem
+If the base ring of `x` is `GF(q^2)`, return the matrix `transpose( map ( y -> y^q, x) )`. An error is returned if the base ring has no even degree.
+"""
+function conjugate_transpose(x::MatElem{T}) where T <: FieldElem
+   iseven(degree(base_ring(x))) || throw(ArgumentError("The base ring must have even degree"))
+   e = div(degree(base_ring(x)),2)
+   return transpose(map(y -> frobenius(y,e),x))
+end
+
+
+# computes a complement for W in V (i.e. a subspace U of V such that V is direct sum of U and W)
+"""
+    complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T})
+Return a complement for `W` in `V`, i.e. a subspace `U` of `V` such that `V` is direct sum of `U` and `W`.
+"""
+function complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T}) where T <: FieldElem
+   @assert issubmodule(V,W) "The second argument is not a subspace of the first one"
+   if dim(W)==0 return sub(V,basis(V)) end
+
+   e = W.map
+
+   H = matrix( vcat([e(g) for g in gens(W)], [zero(V) for i in 1:(dim(V)-dim(W)) ]) )
+   d = dim(W)
+   A_left = identity_matrix(base_ring(V), dim(V))
+   A_right = identity_matrix(base_ring(V), dim(V))
+   for rn in 1:dim(W)     # rn = row number
+      cn = rn    # column number
+      while H[rn,cn]==0 cn+=1 end   # bring on the left the first non-zero entry
+      swap_cols!(H,rn,cn)
+      swap_rows!(A_right,rn,cn)
+      for j in rn+1:dim(W)
+         add_row!(H,H[j,rn]*H[rn,rn]^-1,rn,j)
+         add_column!(A_left,A_left[j,rn]*A_left[rn,rn]^-1,j,rn)
+      end
+   end
+   for j in dim(W)+1:dim(V)  H[j,j]=1  end
+   H = A_left*H*A_right
+   _gens = [V([H[i,j] for j in 1:dim(V)]) for i in dim(W)+1:dim(V) ]
+
+   return sub(V,_gens)
+end
+
+"""
+    permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Int
+Return the permutation matrix over the ring `R` corresponding to `Q`. Here, `Q` must contain exactly once every integer from 1 to some `n`.
+"""
+function permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Int
```

Comment:
I think you want to say `where T <: Integer`.

Reply:
With `Integer` it does not work. It returns the error
> ERROR: MethodError: no method matching permutation_matrix(::FqNmodFiniteField, ::Array{Int64,1})

The same error  is returned if the first line is replaced by
> permutation_matrix(F::Ring, Q::AbstractVector{<:T}) where T <: Integer

Reply:
I think one has to write `T <: Base.Integer`, since `Integer` (=`Oscar.Integer`) means `fmpz` inside `Oscar`.